notifications: Render emoji in desktop notifications.

saubhagya-patel opened the pull request:
Through the changes in this PR, in desktop notifications, we no longer display an emoji's status code. Instead, we show the corresponding Unicode emoji to render it properly in desktop notifications. For custom emojis, we continue to display their name.


<!-- Describe your pull request here.-->

Fixes: #30598.
[CZO thread](https://chat.zulip.org/#narrow/stream/137-feedback/topic/Emoji.20in.20windows.20notification/near/1815298).

**Previous PR Reference:**
- This PR continues the work done in the PR [#30653](https://github.com/zulip/zulip/pull/30653) by @tnmkr.
- There also exists a PR [#31949](https://github.com/zulip/zulip/pull/31949) by @tony-nyagah.
- The developments made in this PR are:
  - Better error handling for custom emoji and for the `String.fromCodePoint()` function.
  - Added tests to verify the logic.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

- Emojis in desktop notifications.

| Before | After |
|--------|-------|
| ![image](https://github.com/user-attachments/assets/8a7e90fc-9c7f-4cb7-827b-7356dc46d0bf) | ![image](https://github.com/user-attachments/assets/d57fa804-491c-4b75-bd0e-c75aaf739912) |
| ![image](https://github.com/user-attachments/assets/e7a7e388-b401-4e39-82af-ad038d4ce68c) | ![image](https://github.com/user-attachments/assets/1dccbae7-2a59-47ba-921b-25a05a912b13) |

- Custom emojis in desktop notifications - (_No changes_).

| Before | After |
|--------|-------|
| ![image](https://github.com/user-attachments/assets/f13021c2-06de-4834-823a-06eeea4ea9a2) | ![image](https://github.com/user-attachments/assets/01108f42-0790-4cef-8de0-102374c55c02) |

- Emojis with [ZWJ sequences](https://www.unicode.org/emoji/charts/emoji-zwj-sequences.html).
  
  | Before | After |
  |--------------|-------------|
  | ![image](https://github.com/user-attachments/assets/87713b08-450b-4d4a-adf7-bfbc8833eef4) | ![image](https://github.com/user-attachments/assets/0d887cb3-b123-4194-a4bd-7b139a3a7644) |
---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/ui_util.ts:
```diff
@@ -22,14 +22,39 @@ export function place_caret_at_end(el: HTMLElement): void {
+
+            // Emojis have a class with the emoji code next to them e.g. ".emoji-1f419".
+            // The code "1f419" represents an octopus.
+            const emoji_class_with_code: string | undefined = $(this).attr("class");
```
This variable name is misleading, since the class attribute is string containing a space-separated list of multiple classes.

timabbott commented on web/src/ui_util.ts:
```diff
@@ -22,14 +22,39 @@ export function place_caret_at_end(el: HTMLElement): void {
+            }
+            const unicode_emoji_code = String.fromCodePoint(hex_emoji_code);
+            return unicode_emoji_code;
         })
```
Can you add tests to `web/tests/ui_util.test.cjs` verifying this logic?

timabbott commented on web/src/ui_util.ts:
```diff
@@ -22,14 +22,39 @@ export function place_caret_at_end(el: HTMLElement): void {
+                Number.isNaN(hex_emoji_code) ||
+                !(hex_emoji_code >= 0 && hex_emoji_code <= 0x10ffff)
+            ) {
+                return $(this).text();
```
I don't think we expect invalid code points to ever happen here. I guess it's fine to have a check for it, but let's do `blueslip.error("Invalid unicode codepoint for emoji", emoji_code)` logging if this happens.

timabbott commented on web/src/ui_util.ts:
```diff
@@ -22,14 +22,39 @@ export function place_caret_at_end(el: HTMLElement): void {
+            const match = regex.exec(emoji_class_with_code);
+            const emoji_code = match?.[1] ?? "";
+
+            const hex_emoji_code = Number.parseInt(emoji_code, 16);
```
I think these variable names are quite confusing. The string you get should be called `emoji_code_hex_string`, because it's a hexadecimal string. The thing you compute here is just an integer, so would most properly be called something like `emoji_code` or `emoji_code_int`.

timabbott commented:
Overall strategy looks fine, posted a handful of comments on polishing the engineering of this to be robust and readable.

tnmkr commented on web/src/ui_util.ts:
```diff
@@ -22,14 +22,39 @@ export function place_caret_at_end(el: HTMLElement): void {
+                return $(this).text();
+            }
+
+            const regex = /emoji-(\w+)/;
```
Looks like this will match only the first codepoint from [zwj sequences](https://www.unicode.org/emoji/charts/emoji-zwj-sequences.html)?

saubhagya-patel replied:
@tnmkr, I am planning to change this regex from `/emoji-(\w+)/` to `/^emoji-([0-9a-fA-F]+)$/` to get a better grip on the hex part. However, I would like a little more context on your above comment, please. Thanks.

tnmkr replied:
I think that approach should be fine, as I cannot recall any specific issues I had with it. But I would still recommend some testing.

I had problems with UX, that i have highlighted in the [#frontened CZO thread](https://chat.zulip.org/#narrow/channel/6-frontend/topic/emoji.20in.20desktop.20notifications/near/2092529).

timabbott commented on web/src/ui_util.ts:
```diff
@@ -22,13 +23,47 @@ export function place_caret_at_end(el: HTMLElement): void {
             }
+
+            const emoji_name = $(this).text().slice(1, -1);
+            const emoji_code_hex_string = emoji.get_emoji_codepoint(emoji_name);
```
Hmm, why are you doing this rather than parsing the CSS classes? The CSS classes are a much stronger guarantee of what codepoint we intend. 

There's a somewhat expectable situation where the name is stale (after an emoji name/codepoint mapping was changed) and thus logic won't work, but it should be very rare.

timabbott commented on web/tests/ui_util.test.cjs:
```diff
@@ -67,3 +67,38 @@ run_test("potentially_collapse_quotes", ({override_rewire}) => {
+    const emoji_name_to_codepoint = {
+        octopus: "1f419",
+        "man cook": "1f468-200d-1f373",
+        "invalid emoji": "invalid",
```
I don't think you need this invalid line? It should be just `undefined` for a value not present here.

saubhagya-patel replied:
Ooh, my bad. This is fixed in the subsequent push.

saubhagya-patel commented:
Following the review received, I have updated the PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/33407#discussion_r1964633547) comment, the `ui_util.replace_emoji_name_with_emoji_unicode()` function is updated to extract the emoji code from the class using the `ui_util.extract_emoji_code_from_class()` function being added as stated [here](https://github.com/zulip/zulip/pull/33407#discussion_r1956524371).

- Updated the tests following [this](https://github.com/zulip/zulip/pull/33407#discussion_r1964634234) comment.

- Visual changes: _None_.

`Tim Abbott` , I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

I am not sure about the failing CI though, will update this as soon as I am able to debug it.

timabbott commented:
The CI failures were an issue with a third-party dependency, not an issue for this PR.

But your tests are completely useless -- they don't call any of the real code. This seems better for the actual function:

``` diff
diff --git a/web/src/ui_util.ts b/web/src/ui_util.ts
index f5578f3d12..f1d3e25e16 100644
--- a/web/src/ui_util.ts
+++ b/web/src/ui_util.ts
@@ -22,9 +22,9 @@ export function place_caret_at_end(el: HTMLElement): void {
     }
 }
 
-function extract_emoji_code_from_class(emoji_class_string: string | undefined): string {
+function extract_emoji_code_from_class(emoji_class_string: string | undefined): string | undefined {
     if (emoji_class_string === undefined) {
-        return "";
+        return undefined;
     }
 
     const classes = emoji_class_string.split(/\s+/);
@@ -33,10 +33,10 @@ function extract_emoji_code_from_class(emoji_class_string: string | undefined):
     for (const cls of classes) {
         const match = regex.exec(cls);
         if (match) {
-            return match?.[1] ?? "";
+            return match?.[1] ?? undefined;
         }
     }
-    return "";
+    return undefined;
 }
 
 export function replace_emoji_name_with_emoji_unicode($element: JQuery): void {
@@ -51,7 +51,7 @@ export function replace_emoji_name_with_emoji_unicode($element: JQuery): void {
 
             const emoji_class_string = $(this).attr("class");
             const emoji_code_hex_string = extract_emoji_code_from_class(emoji_class_string);
-            if (emoji_code_hex_string === "") {
+            if (emoji_code_hex_string === undefined) {
                 return $(this).text();
             }
 
diff --git a/web/tests/ui_util.test.cjs b/web/tests/ui_util.test.cjs
index bc7774107b..bedda2ca34 100644
--- a/web/tests/ui_util.test.cjs
+++ b/web/tests/ui_util.test.cjs
@@ -95,6 +95,5 @@ run_test("replace_emoji_name_with_emoji_unicode", () => {
     const man_cook_emoji = "👨‍🍳";
     assert.equal(man_cook_emoji, get_emoji_unicode(emoji_name_to_codepoint["man cook"]));
 
-    const undefined_emoji = "";
-    assert.equal(undefined_emoji, get_emoji_unicode(emoji_name_to_codepoint["undefined emoji"]));
+    assert.equal("", get_emoji_unicode(emoji_name_to_codepoint["undefined emoji"]));
 });
```

timabbott commented:
I reworked this to have an actually useful test:

``` diff
diff --git a/web/src/message_notifications.ts b/web/src/message_notifications.ts
index 2b98e6cfe2..f435f272a3 100644
--- a/web/src/message_notifications.ts
+++ b/web/src/message_notifications.ts
@@ -40,7 +40,7 @@ function get_notification_content(message: Message | TestNotificationMessage): s
     let content;
     // Convert the content to plain text, replacing emoji with their alt text
     const $content = $("<div>").html(message.content);
-    ui_util.replace_emoji_name_with_emoji_unicode($content);
+    ui_util.convert_unicode_eligible_emoji_to_unicode($content);
     ui_util.change_katex_to_raw_latex($content);
     ui_util.potentially_collapse_quotes($content);
     spoilers.hide_spoilers_in_notification($content);
diff --git a/web/src/ui_util.ts b/web/src/ui_util.ts
index f1d3e25e16..3970ab6752 100644
--- a/web/src/ui_util.ts
+++ b/web/src/ui_util.ts
@@ -39,44 +39,43 @@ function extract_emoji_code_from_class(emoji_class_string: string | undefined):
     return undefined;
 }
 
-export function replace_emoji_name_with_emoji_unicode($element: JQuery): void {
-    $element
-        .find(".emoji")
-        .text(function () {
-            // This is a custom emoji, we do not have corresponding emoji
-            // unicode for these so we return original markdown.
-            if ($(this).is("img")) {
-                return $(this).attr("alt") ?? "";
-            }
+export function convert_emoji_element_to_unicode($emoji_elt): string {
+    // This is a custom emoji, we do not have corresponding emoji
+    // unicode for these so we return original markdown.
+    if ($emoji_elt.is("img")) {
+        return $emoji_elt.attr("alt") ?? "";
+    }
 
-            const emoji_class_string = $(this).attr("class");
-            const emoji_code_hex_string = extract_emoji_code_from_class(emoji_class_string);
-            if (emoji_code_hex_string === undefined) {
-                return $(this).text();
+    const emoji_class_string = $emoji_elt.attr("class");
+    const emoji_code_hex_string = extract_emoji_code_from_class(emoji_class_string);
+    if (emoji_code_hex_string === undefined) {
+        return $emoji_elt.text();
+    }
+
+    const emoji_code_parts = emoji_code_hex_string.split("-");
+    const emoji_unicode = emoji_code_parts
+        .map((emoji_code) => {
+            const emoji_code_int = Number.parseInt(emoji_code, 16);
+            // Validate the parameter passed to String.fromCodePoint() (here, emoji_code_int).
+            // "An integer between 0 and 0x10FFFF (inclusive) representing a Unicode code point."
+            // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/fromCodePoint
+            // for details.
+            if (
+                Number.isNaN(emoji_code_int) ||
+                    !(emoji_code_int >= 0 && emoji_code_int <= 0x10ffff)
+            ) {
+                blueslip.error("Invalid unicode codepoint for emoji", {emoji_code_int});
+                return $emoji_elt.text();
             }
+            return String.fromCodePoint(emoji_code_int);
+        }).join("");
+    return emoji_unicode;
+}
 
-            const emoji_code_parts = emoji_code_hex_string.split("-");
-            const emoji_unicode = emoji_code_parts
-                .map((emoji_code) => {
-                    const emoji_code_int = Number.parseInt(emoji_code, 16);
-                    // Validate the parameter passed to String.fromCodePoint() (here, emoji_code_int).
-                    // "An integer between 0 and 0x10FFFF (inclusive) representing a Unicode code point."
-                    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/fromCodePoint
-                    // for details.
-                    if (
-                        Number.isNaN(emoji_code_int) ||
-                        !(emoji_code_int >= 0 && emoji_code_int <= 0x10ffff)
-                    ) {
-                        blueslip.error("Invalid unicode codepoint for emoji", {emoji_code_int});
-                        return $(this).text();
-                    }
-                    return String.fromCodePoint(emoji_code_int);
-                })
-                .join("");
-            return emoji_unicode;
-        })
-        .contents()
-        .unwrap();
+export function convert_unicode_eligible_emoji_to_unicode($element: JQuery): void {
+    return $element.find(".emoji").text(() => {
+        return convert_emoji_element_to_unicode(this);
+    }).contents().unwrap();
 }
 
 export function change_katex_to_raw_latex($element: JQuery): void {
diff --git a/web/tests/ui_util.test.cjs b/web/tests/ui_util.test.cjs
index bedda2ca34..46cd20b068 100644
--- a/web/tests/ui_util.test.cjs
+++ b/web/tests/ui_util.test.cjs
@@ -68,32 +68,18 @@ run_test("potentially_collapse_quotes", ({override_rewire}) => {
     );
 });
 
-function get_emoji_unicode(emoji_code_hex_string) {
-    if (!emoji_code_hex_string) {
-        return "";
-    }
-    const emoji_code_parts = emoji_code_hex_string.split("-");
-    const emoji_unicode = emoji_code_parts
-        .map((emoji_code) => {
-            const emoji_code_int = Number.parseInt(emoji_code, 16);
-            return String.fromCodePoint(emoji_code_int);
-        })
-        .join("");
-
-    return emoji_unicode;
-}
-
 run_test("replace_emoji_name_with_emoji_unicode", () => {
     const emoji_name_to_codepoint = {
         octopus: "1f419",
         "man cook": "1f468-200d-1f373",
     };
+    let $emoji = $.create("span").attr("class", "emoji emoji-1f419");
+    $emoji.is = () => false;
 
     const octopus_emoji = "🐙";
-    assert.equal(octopus_emoji, get_emoji_unicode(emoji_name_to_codepoint.octopus));
+    assert.equal(octopus_emoji, ui_util.convert_emoji_element_to_unicode($emoji));
 
+    $emoji.attr("class", "emoji emoji-1f468-200d-1f373");
     const man_cook_emoji = "👨‍🍳";
-    assert.equal(man_cook_emoji, get_emoji_unicode(emoji_name_to_codepoint["man cook"]));
-
-    assert.equal("", get_emoji_unicode(emoji_name_to_codepoint["undefined emoji"]));
+    assert.equal(man_cook_emoji, ui_util.convert_emoji_element_to_unicode($emoji));
 });
```

timabbott commented:
I cleaned the actual implementation up, but there's some linter/typing errors that I don't have time to fix before stopping for the day. @saubhagya-patel please fix those and let me know when that's done.

saubhagya-patel commented:
@timabbott, the lint errors have been resolved.

Thanks for improving the code and tests.
It’s a helpful lesson for me on how to write comprehensive tests and maintain clean code.

timabbott commented:
Merged, thanks @saubhagya-patel and @tnmkr!
